### PR TITLE
Make UriBuilder.Query setter idempotent

### LINKED ISSUE #3646: UriBuilder.Query property setter is not idempotent
The `UriBuilder.Query` property setter prepends the supplied value with a `?` character, even if the supplied value already starts with `?`.

https://github.com/dotnet/corefx/blob/41e203011152581a6c65bb81ac44ec037140c1bb/src/System.Private.Uri/src/System/UriBuilder.cs#L277

This means that code calling the `Query` setter has to check to ensure the value it's passing doesn't already have the question mark even when hoping to simply append to the end of the query.

Here's example code that suffers from the issues this surfaces:

``` csharp
var uri = new UriBuilder("http://localhost/page?param1=foo");

if (someOption) {
    uri.Query += "&someOption=true";
}
```

Expected value of uri.Query:
`?param1=foo&someOption=true`

Actual value of uri.Query:
`??param1=foo&someOption=true`

### PR DESCRIPTION
**Pull request by jeffhandley**
Fixes #3646 in a way that is hopefully determined to be backwards compatible.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @cipop, @davidsh, @SidharthNabar

File: src/System.Private.Uri/src/System/UriBuilder.cs [hunk 1]
```diff
@@ -276,7 +276,7 @@ public string Query
                 }
                 if (value.Length > 0)
                 {
-                    value = '?' + value;
+                    value = value.StartsWith("?") ? value : '?' + value;
```
Comment (stephentoub) [inline src/System.Private.Uri/src/System/UriBuilder.cs]: Assuming this is a fix that's desired, not a breaking change we care about, etc. I'd think it'd be a bit better to write it as:
```C#
if (value.Length > 0 && value[0] != '?')
{
    value = "?" + value;
}
```

Comment (jasonwilliams200OK) [inline src/System.Private.Uri/src/System/UriBuilder.cs]: Should `value` be `Trim()`med before entering this block or `TrimStart()`?

**Comment by davidsh** [comment]
Changing this class is potentially app-compat breaking...so we'll need to study these changes carefully.

**Comment by jeffhandley** [comment]
@stephentoub Okay, I can update that.

For backwards compat, the only scenario I can think if that would change behavior is `uri.Query = "?foo"`, which would have previously resulted in a double question mark, but now wouldn't.  That is the fix, and I can't think of how having a double question mark would have ever been desired.  Code that was working around this behavior should continue working as is since we'll still pretend the question mayo if it isn't supplied.

**Comment by jeffhandley** [comment]
@stephentoub I updated the Query setter code per your suggestion.  Good catch on that.

**Comment by CIPop** [comment]
@terrajobst @weshaggard @ericstj : While this is a bug, changing the code in corefx will create a different behavior for UriBuilder.

What is your opinion on this?

**Comment by CIPop** [comment]
@jeffhandley Please also add the test you mentioned in #3646.

**Comment by davidsh** [comment]
>While this is a bug, changing the code in corefx will create a different behavior for UriBuilder.

And specifically it will be different from .NET Framework 4.6 (Desktop)

**Comment by jeffhandley** [comment]
@CIPop I don't see any existing tests for UriBuilder (or anything else under that area).
https://github.com/dotnet/corefx/tree/master/src/System.Private.Uri

Please let me know how you'd like me to proceed in adding the tests; I'd be glad to write some more UriBuilder tests but I can't commit to spinning up a new test project for the area.

**Comment by davidsh** [comment]
Those classes officially live in the System.Runtime contract.  So, I would expect the tests to live under that node.

**Comment by davidsh** [comment]
https://github.com/dotnet/corefx/tree/master/src/System.Runtime/tests/System has Uri related tests.

**Comment by jeffhandley** [comment]
@davidsh Ah, perfect!  I'll add some UriBuilder tests in there.  Thanks!

**Comment by CIPop** [comment]
@jeffhandley you're right :( none of the tests got ported from our previous system yet.

That is not the location you would add the tests. Instead the tests should be in System.Private.Uri\tests\.

**Comment by davidsh** [comment]
@stephentoub, @weshaggard 

>That is not the location you would add the tests. Instead the tests should be in System.Private.Uri\tests.

I thought tests should be in the "contract" library source folder tree and not in the System.Private.* tree.  If not, are we consistent about this in CoreFx?

**Comment by jeffhandley** [comment]
@davidsh @CIPop Let me know which way you want to go on the tests.  If there's an existing set of tests that we could have ported over to seed from, that would be good.

Here's what I wrote up to exercise the Query getter/setter.

``` csharp
    [Fact]
    public static void TestQuery()
    {
        UriBuilder uriBuilder = new UriBuilder(@"http://foo/bar/baz?date=today");

        string q;

        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"?date=today");

        uriBuilder.Query = @"?date=yesterday";
        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"?date=yesterday");

        uriBuilder.Query = @"date=tomorrow";
        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"?date=tomorrow");

        uriBuilder.Query += @"&place=redmond";
        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"?date=tomorrow&place=redmond");

        uriBuilder.Query = null;
        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"");

        uriBuilder.Query = "";
        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"");

        uriBuilder.Query = "?";
        q = uriBuilder.Query;
        Assert.Equal<String>(q, @"?");
    }
```

**Comment by CIPop** [comment]
Please place the tests within System.Private.Networking.
Also, please add a comment specifying that this test is checking for different behavior compared to .Net Desktop.

**Comment by davidsh** [comment]
>Please place the tests within System.Private.Networking.

Shouldn't this be `System.Private.Uri` and not `System.Private.Networking` ?

**Comment by CIPop** [comment]
@davidsh Thanks! Corrected above.

**Comment by jasonwilliams200OK** [comment]
@CIPop, it is not clear from the docs, what is `System.Private` namespace. Is it transient namespace which will eventually fade away at some point in favor of full publicly consumable package?

@davidsh, you mentioned at https://github.com/dotnet/corefx/issues/1745 that System.Private.Uri tests will be ported to GitHub at some point. Is this PR actually awaiting that (so @jeffhandley can add tests in the upcoming `System.Private.Uri.Tests`  project)?

**Comment by jeffhandley** [comment]
Moved the test file over to a new test project under System.Private.Uri, but I don't presently have a machine set up where I can even run a build, so this will need some follow-up attention (can someone help out?).

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by jasonwilliams200OK** [comment]
@jeffhandley, seems like `Properties/AssemblyInfo.cs` is required to execute the test on CI like [this](https://github.com/dotnet/corefx/blob/c02d33b/src/System.Diagnostics.Process/tests/Properties/AssemblyInfo.cs) probably with `[assembly: AssemblyTitle("System.Private")]`.

**Comment by stephentoub** [comment]
@davidsh or @cipop, what's the plan for this?  It's been open since October.  Can we decide to either not take the change and close it, or take the change and get everything fixed up to merge it?

**Comment by davidsh** [comment]
No all feedback on this PR was addressed yet.

>From: @CIPop 
Also, please add a comment specifying that this test is checking for different behavior compared to .Net Desktop. 

Not addressed yet.

>Asked questions of @terrajobst @weshaggard @ericstj : While this is a bug, changing the code in corefx will create a different behavior for UriBuilder. What is your opinion on this?

Not addressed yet.  We have not heard back from @terrajobst @weshaggard or @ericstj 

If we can get traction on those items, then this PR can be cleaned up and prepared for merge.

**Comment by stephentoub** [comment]
> Asked questions of @terrajobst @weshaggard @ericstj : While this is a bug, changing the code in corefx will create a different behavior for UriBuilder. What is your opinion on this?

@weshaggard, could you weigh in on this one? Thanks!

**Comment by weshaggard** [comment]
I don't honestly have enough context to know what bad can come from the change. It personally looks pretty harmless to me and we should consider it for desktop as well but ultimately @CIPop and @davidsh have more experience in the URI space and know the sort of troubles people run into so I think they are in a better position to make a call on this.

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by davidsh** [comment]
@jeffhandley There are compile errors here with the CSPROJ you added:

>Compile Error

error CS0246: The type or namespace name 'AssemblyTitleAttribute' could not be found (are you missing a using directive or an assembly reference?) [d:\j\workspace\windows_nt_re---5067c899\src\System.Private.Uri\tests\System.Private.Uri.CoreCLR.Tests.csproj]
Indication 1 

I recommend you don't use Visual Studio to create the CSPROJ. You should simply copy a known working CSPROJ from one of our other test projects (i.e. System.Net.Http/tests/FunctionalTests/*.CSPROJ.

Can you fix up this PR so that it can be validated? Thx.

**Comment by jeffhandley** [comment]
@vcsjones offered to help me out on this.  I don't presently have a machine set up where I could work on this stuff.

**Comment by vcsjones** [comment]
I've cherry-picked @jeffhandley's changes, but I have a few questions to get the ball rolling again, hopefully quick.

1. Since this PR was opened tests were added for `UriBuilder` to System.Runtime in 3fbf632. Is the goal to still test this in System.Private.Uri? There are tests that explicitly test the old behavior. If System.Private.Uri is the correct approach, how do I reconcile with the tests in System.Runtime?
1. What is the best way to continue this PR? Close this one and have me submit another one for #3646, or something different?

**Comment by jasonwilliams200OK** [comment]
My $0.002: you can send a PR to @jeffhandley's fork and your commits will automatically emerge as soon as Jeff merge it, retaining the code review, conversation history etc. :)

**Comment by davidsh** [comment]
>Since this PR was opened tests were added for UriBuilder to System.Runtime in 3fbf632. Is the goal to still test this in System.Private.Uri? There are tests that explicitly test the old behavior. If System.Private.Uri is the correct approach, how do I reconcile with the tests in System.Runtime?

Yes. We wanted the tests under System.Private.Uri to help consolidate all Uri* class tests.  It is unfortunate that we allowed UriBuilder tests to be under System.Runtime library folder.  We will move them at some point.

>What is the best way to continue this PR? Close this one and have me submit another one for #3646, or something different?

Yes, close this PR and start a new one. Provide a reference link back.

**Comment by stephentoub** [comment]
Replaced by https://github.com/dotnet/corefx/pull/6926

**Pull request closed without merging**